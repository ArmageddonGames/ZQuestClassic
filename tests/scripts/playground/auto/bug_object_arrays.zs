#include "std.zh"
#include "auto/test_runner.zs"

generic script bug_object_arrays
{
	void run()
	{
		Test::Init();

		Waitframe();

		itemsprite item = Screen->CreateItem(I_HEARTCONTAINER);
		Test::Assert(item);

		// TODO: fix this bug.
		// item[] items = {item};
		// Test::Assert(items);
		// Test::Assert(items[0]);
		// Test::AssertEqual(items[0], item);

		// int[] arr = {1337};
		// Test::AssertEqual(arr[0], 1337);
	}
}
