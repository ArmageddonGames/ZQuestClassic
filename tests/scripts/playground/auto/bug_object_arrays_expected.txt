stderr:

Compiling 'bug_object_arrays.zs'
Pass 1: Parsing
Pass 2: Preprocessing
Pass 3: Registration
Pass 4: Analyzing Code
Pass 5: Checking code paths
Pass 6: Generating object code
Pass 7: Assembling
Success!
Compile finished with exit code '0' (success)

stdout:

{
  "success": true,
  "diagnostics": [],
  "metadata": {
    "currentFileSymbols": 3,
    "symbols": 10,
    "identifiers": 10,
    "elided": true
  }
}

zasm:

PUSHARGSV 0 0.0006; void run() Body Start
SETR D4 SP2
ALLOCATEMEMV D2 100 0
STORE D2 0.0004
LOAD D2 0.0004; InlineFunc[void Game::GetSaveName(char32[])] Params
GETSAVENAME D2; InlineFunc[void Game::GetSaveName(char32[])] Body
ALLOCATEMEMV D2 19 0
STORE D2 0.0003
WRITEPODSTRING D2 "[Test] Running %s\n"
LOAD D2 0.0003; InlineFunc[void printf(char32[])] Params Start
PUSHR D2
LOAD D2 0.0004
PUSHVARGR D2; InlineFunc[void printf(char32[])] Params End
PRINTFVARG; InlineFunc[void printf(char32[])] Body Start
POP D5; InlineFunc[void printf(char32[])] Body End
LOAD D3 0.0003
DEALLOCATEMEMR D3
LOAD D2 0.0004; InlineFunc[int Game::GetGenericScript(char32[])] Params
GETGENERICSCRIPT D2; InlineFunc[int Game::GetGenericScript(char32[])] Body
STORE D2 0.0002
LOAD D2 0.0002; if() #448 Test
COMPAREV D2 -1
SETCMP D2 I==
COMPAREV D2 0
GOTOCMP 36 ==; Test 'if'
ALLOCATEMEMV D2 53 0; if() #448 Body Start
STORE D2 0.0001
WRITEPODSTRING D2 "[Test] failed to start - could not find test script\n"
LOAD D2 0.0001; InlineFunc[void printf(char32[])] Params Start
PUSHR D2; InlineFunc[void printf(char32[])] Params End
PRINTFVARG; InlineFunc[void printf(char32[])] Body Start
POP D5; InlineFunc[void printf(char32[])] Body End
LOAD D3 0.0001
DEALLOCATEMEMR D3
GAMEEXIT; InlineFunc[void Game::Exit()] Body AND if() #448 Body End
LOAD D2 0.0002; InlineFunc[genericdata Game::LoadGenericData(int)] Params
LOADGENERICDATA D2; InlineFunc[genericdata Game::LoadGenericData(int)] Body Start
SETR D2 REFGENERICDATA; InlineFunc[genericdata Game::LoadGenericData(int)] Body End
STORE D2 0.0001
SETV D2 1
PUSHR D2
LOAD D2 0.0001
POP D3
PUSHR D2
PUSHR D3
POP D2
CASTBOOLI D2
POP REFGENERICDATA
SETR GENDATARUNNING D2
LOAD D2 0.0001; while() #2 Test
PUSHR D2
POP REFGENERICDATA
SETR D2 GENDATARUNNING
COMPAREV D2 0
GOTOCMP 58 ==; Test 'while'
WAITFRAME; InlineFunc[void Waitframe()] Body AND while() #2 Body
GOTO 50; while() #2 End
ALLOCATEMEMV D2 13 0
STORE D2 0
WRITEPODSTRING D2 "[Test] done\n"
LOAD D2 0; InlineFunc[void printf(char32[])] Params Start
PUSHR D2; InlineFunc[void printf(char32[])] Params End
PRINTFVARG; InlineFunc[void printf(char32[])] Body Start
POP D5; InlineFunc[void printf(char32[])] Body End
LOAD D3 0
DEALLOCATEMEMR D3
GAMEEXIT; InlineFunc[void Game::Exit()] Body
LOAD D3 0.0004
DEALLOCATEMEMR D3; [Opt:NeverRet] AND void run() Body End


PUSHR REFGENERICDATA; void run() Body Start
PUSHV 0
SETR D4 SP2
PUSHR D4
PUSHV 5; UsrFunc[void Init(int, int)] Params Start
PUSHV 15; UsrFunc[void Init(int, int)] Params End
CALLFUNC 0.0052; UsrFunc[void Init(int, int)] Call
POP D4
WAITFRAME; InlineFunc[void Waitframe()] Body
SETV D2 28; InlineFunc[itemsprite screendata::CreateItem(int)] Params
CREATEITEMR D2; InlineFunc[itemsprite screendata::CreateItem(int)] Body Start
SETR D2 REFITEM; InlineFunc[itemsprite screendata::CreateItem(int)] Body End
STORE D2 0
PUSHR D4
LOAD D2 0; UsrFunc[void Assert(itemsprite)] Params Start
PUSHR D2; UsrFunc[void Assert(itemsprite)] Params End
CALLFUNC 0.0086; UsrFunc[void Assert(itemsprite)] Call
POP D4
POPARGS D5 0.0002
QUIT; void run() Body End
PUSHV 0; Func[void Fail(char32[])] Body Start
SETR D4 SP2
LOAD D2 0.0001; InlineFunc[int strlen(char32[])] Params Start AND if() #160 Test
SETR D3 D2; InlineFunc[int strlen(char32[])] Params End
STRINGLENGTH D2 D3; InlineFunc[int strlen(char32[])] Body
COMPAREV D2 0
SETCMP D2 I>
COMPAREV D2 0
GOTOCMP 42 ==; Test 'if'
ALLOCATEMEMV D2 24 0; if() #160 Body Start
STORE D2 0
WRITEPODSTRING D2 "[Test] failed test: %s\n"
LOAD D2 0; InlineFunc[void printf(char32[])] Params Start
PUSHR D2
LOAD D2 0.0001
PUSHVARGR D2; InlineFunc[void printf(char32[])] Params End
PRINTFVARG; InlineFunc[void printf(char32[])] Body Start
POP D5; InlineFunc[void printf(char32[])] Body End
LOAD D3 0
DEALLOCATEMEMR D3
GOTO 51; if() #160 Body End
ALLOCATEMEMV D2 20 0; if() #160 Else Start
STORE D2 0
WRITEPODSTRING D2 "[Test] failed test\n"
LOAD D2 0; InlineFunc[void printf(char32[])] Params Start
PUSHR D2; InlineFunc[void printf(char32[])] Params End
PRINTFVARG; InlineFunc[void printf(char32[])] Body Start
POP D5; InlineFunc[void printf(char32[])] Body End
LOAD D3 0
DEALLOCATEMEMR D3; if() #160 Else End
GAMEEXIT; InlineFunc[void Game::Exit()] Body AND [Opt:NeverRet] AND Func[void Fail(char32[])] Body End
PUSHV 0; Func[void Init(int, int)] Body Start
SETR D4 SP2
ALLOCATEMEMV D2 35 0
STORE D2 0
WRITEPODSTRING D2 "[Test] started: dmap %d screen %d\n"
LOAD D2 0; InlineFunc[void printf(char32[])] Params Start
PUSHR D2
LOAD D2 0.0002
PUSHVARGR D2
LOAD D2 0.0001
PUSHVARGR D2; InlineFunc[void printf(char32[])] Params End
PRINTFVARG; InlineFunc[void printf(char32[])] Body Start
POP D5; InlineFunc[void printf(char32[])] Body End
LOAD D3 0
DEALLOCATEMEMR D3
LOAD D2 0.0002; if() #447 Test
COMPAREV D2 5
GOTOCMP 75 !=
LOAD D2 0.0001
COMPAREV D2 15
GOTOCMP 75 !=
SETV D2 0
GOTO 76
SETV D2 1
COMPAREV D2 0
GOTOCMP 84 ==; Test 'if'
LOAD D2 0.0002; InlineFunc[void player::Warp(int, int)] Params Start AND if() #447 Body Start
PUSHR D2
LOAD D2 0.0001; InlineFunc[void player::Warp(int, int)] Params End
POP D3; InlineFunc[void player::Warp(int, int)] Body Start
POP D6
WARPR D3 D2; InlineFunc[void player::Warp(int, int)] Body End AND if() #447 Body End
POPARGS D5 0.0003
RETURNFUNC; Func[void Init(int, int)] Body End
PUSHV 0; Func[void Assert(itemsprite)] Body Start
SETR D4 SP2
LOAD D2 0.0001; unless() #449 Test
COMPAREV D2 0
GOTOCMP 99 !=; Test 'unless'
ALLOCATEMEMV D2 1 0; unless() #449 Body Start
STORE D2 0
WRITEPODSTRING D2 ""
LOAD D2 0; UsrFunc[void Fail(char32[])] Params Start
PUSHR D2; UsrFunc[void Fail(char32[])] Params End
CALLFUNC 0.0021; UsrFunc[void Fail(char32[])] Call AND [Opt:NeverRet]
LOAD D3 0
DEALLOCATEMEMR D3; unless() #449 Body End
POPARGS D5 0.0002
RETURNFUNC; Func[void Assert(itemsprite)] Body End


QUIT; void run() Body
